### PR TITLE
[feature](nereids) Date add sub related functions

### PR DESCRIPTION
**Pull request by englefly**
# Proposed changes

## date_add series
- DATE_ADD
- DAYS_ADD
- ADDDATE
- TIMESTAMPADD

## date_sub series
- DATE_SUB
- DAYS_SUB
- SUBDATE

## NOTE
1. For DAYS_XXX, time unit is omissible, by default the time unit is DAY
2. no TIMESTAMPSUB

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
5. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
6. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
@morrySnow PTAL

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.List;
+
+/**
+ * ScalarFunction 'days_add'.
+ */
+public class DaysAdd extends ScalarFunction implements BinaryExpression, ExplicitlyCastableSignature, AlwaysNullable {
+    public static final List<FunctionSignature> SIGNATURES = ImmutableList.of(
+            FunctionSignature.ret(DateTimeType.INSTANCE).args(DateTimeType.INSTANCE, IntegerType.INSTANCE),
+            FunctionSignature.ret(DateTimeType.INSTANCE).args(DateTimeV2Type.INSTANCE, IntegerType.INSTANCE),
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java:42]: if arg is DateTimeV2Type, the return type should be DateTimeV2Type too

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            FunctionSignature.ret(DateTimeType.INSTANCE).args(DateV2Type.INSTANCE, IntegerType.INSTANCE)
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java:43]: arg could be date, right?

File: regression-test/suites/nereids_syntax_p0/test_date_add.groovy [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+suite("test_date_add") {
+    sql "set enable_nereids_planner=true"
+    sql "set enable_fallback_to_original_planner=false"
+
+    
+    qt_select """
+    SELECT TIMESTAMPADD('2003-02-01', interval 2 year)
```
Comment (morrySnow) [inline regression-test/suites/nereids_syntax_p0/test_date_add.groovy:24]: we need to test four types: `date`, `datev2`, `datetime`, `datetimev2`

Comment (englefly) [reply regression-test/suites/nereids_syntax_p0/test_date_add.groovy:24]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java:43]: yes, I missed them

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/scalar/DaysAdd.java:42]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/ExpressionTrait.java [hunk 4]
```diff
@@ -48,6 +48,6 @@ default DataType getDataType() throws UnboundException {
     }
 
     default String toSql() throws UnboundException {
-        throw new UnboundException("sql");
+        throw new UnboundException(this.toString());
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/functions/ExpressionTrait.java:51]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java [hunk 5]
```diff
@@ -65,7 +65,6 @@ public enum Operator {
         BITXOR("^", "bitxor", OperatorPosition.BINARY_INFIX, TExprOpcode.BITXOR),
         BITNOT("~", "bitnot", OperatorPosition.UNARY_PREFIX, TExprOpcode.BITNOT),
         FACTORIAL("!", "factorial", OperatorPosition.UNARY_POSTFIX, TExprOpcode.FACTORIAL);
-
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ArithmeticExpr.java:68]: done

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.g4 [hunk 6]
```diff
@@ -143,13 +144,16 @@ CURRENT_TIME: 'CURRENT_TIME';
 CURRENT_TIMESTAMP: 'CURRENT_TIMESTAMP';
 CURRENT_USER: 'CURRENT_USER';
 DAY: 'DAY';
+DAYS_ADD: 'DAYS_ADD';
+DAYS_SUB: 'DAYS_SUB';
 DATA: 'DATA';
 DATABASE: 'DATABASE';
 DATABASES: 'DATABASES';
 DATE: 'DATE';
 DATEADD: 'DATEADD';
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.g4:153]: no DATEADD and DATESUB

File: fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.g4 [hunk 7]
```diff
@@ -90,6 +90,7 @@ RIGHT_BRACKET: ']';
 //============================
 //--DORIS-KEYWORD-LIST-START
 ADD: 'ADD';
+ADDDATE:'ADDDATE';
```
[the remark this answers is not in the document]
Comment (englefly) [reply fe/fe-core/src/main/antlr4/org/apache/doris/nereids/DorisLexer.g4:93]: done

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 36.31 seconds
 load time: 486 seconds
 storage size: 17123356305 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221202102209_clickbench_pr_56861.html

**Review by morrySnow — APPROVED** [review]

**Pull request merged**